src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -325,6 +356,459 @@ CreateInsertSelectRouterPlan(Query *originalQuery,
+		{
+			RestrictInfo *rinfo = (RestrictInfo *) lfirst(restrictionInfoList);
+			OpExpr *restrictionOpExpr = NULL;
+			Expr *leftExpr = NULL;
+			Expr *rightExpr = NULL;
+			Var *leftVar = NULL;
+			Var *rightVar = NULL;
+			Expr *restrictionClause = rinfo->clause;
+
+			if (!IsA(restrictionClause, OpExpr))
+			{
+				continue;
+			}
+
+			restrictionOpExpr = (OpExpr *) restrictionClause;
+			if (list_length(restrictionOpExpr->args) != 2)
+			{
+				continue;
+			}
+			if (!OperatorImplementsEquality(restrictionOpExpr->opno))
+			{
+				continue;
+			}
+
+			leftExpr = linitial(restrictionOpExpr->args);
+			rightExpr = lsecond(restrictionOpExpr->args);
+			if (!(IsA(leftExpr, Var) && IsA(rightExpr, Var)))
+			{
+				continue;
+			}
+
+			leftVar = (Var *) leftExpr;
+			rightVar = (Var *) rightExpr;
+
+			/* we could (probably) safely do that since we disallowed ANTI JOINs ? */
+			AddToVarEquivalenceClass(joinRestriction->plannerInfo, leftVar,
+									 &varEquivalance);
+			AddToVarEquivalenceClass(joinRestriction->plannerInfo, rightVar,
+									 &varEquivalance);
+
+			varEquivalences = lappend(varEquivalences, varEquivalance);
+		}
+	}
+
+	return varEquivalences;
+}
+
+
+/*
+ * ListConcatUniqueEquivalancePartitionKeys get the common equivalence class and
+ * a new equivalence class. It iterates on the members of the new class and adds
+ * the partition key members which are equivalant to the common class.
+ */
+static void
+ListConcatUniqueEquivalantPartitionKeys(VarEquivalenceClass **commonEqClass,
+										List *newEqClass)
+{
+	ListCell *newEqClassCell = NULL;
+
+	foreach(newEqClassCell, newEqClass)
+	{
+		VarEquivalenceClassMember *newEqMember = lfirst(newEqClassCell);
+		Var *relationPartitionKey = PartitionKey(newEqMember->relationId);
+
+		if (!relationPartitionKey)
+		{
+			continue;
+		}
+
+		if (relationPartitionKey->varattno != newEqMember->varattno)
+		{
+			continue;
+		}
+
+		if (VarClassMemberEqualsToVarClass(newEqMember, *commonEqClass))
+		{
+			continue;
+		}
+
+		(*commonEqClass)->equivalentVars = lappend((*commonEqClass)->equivalentVars,
+												   newEqMember);
+	}
+}
+
+
+/*
+ * TODO: (Andres asked) could the rte_relation that we're adding the eq.
+ * be part of ANTI-JOIN?
+ *
+ * AddToVarEquivalenceClass add the varToBeAdded to the given varEquivalanceClass.
+ * The function add the varToBeAdd if it belongs to a relation. If the varToBeAdded
+ * belongs to a subquery, the function recursively calls itself until it finds the
+ * relation that the var actually belongs to.
+ */
+static void
+AddToVarEquivalenceClass(PlannerInfo *root, Var *varToBeAdded,
+						 VarEquivalenceClass **varEquivalanceClass)
+{
+	RangeTblEntry *rte = root->simple_rte_array[varToBeAdded->varno];
+
+	if (rte->rtekind == RTE_RELATION)
+	{
+		VarEquivalenceClassMember *varEqMember = palloc0(
+			sizeof(VarEquivalenceClassMember));
+		varEqMember->varattno = varToBeAdded->varattno;
+		varEqMember->varno = varToBeAdded->varno;
```

Comment:
As far as I can see, per appearance of a relation, Postgres generates an RTE and assigns a unique `RTEIndex`. 
Vars of those relations are assigned the unique rteIndex.
(Postgres also generates RTEs for subqueries etc. but that is not in the scope of this discussion.)
 
For example:

```SQL
INSERT INTO agg_events
            (user_id, value_4_agg)
SELECT
  outer_most.id, max(outer_most.value)
FROM
(
  SELECT f2.id as id, f2.v4 as value FROM
    (SELECT
          id
      FROM   (SELECT reference_table.user_id      AS id
               FROM   raw_events_first,
                      reference_table
            WHERE  raw_events_first.user_id = reference_table.user_id ) AS foo) as f
  INNER JOIN
    (SELECT v4,
          v1,
          id
    FROM   (SELECT SUM(raw_events_second.value_4) AS v4,
               SUM(raw_events_first.value_1) AS v1,
               raw_events_second.user_id      AS id
            FROM   raw_events_first,
                    raw_events_second
            WHERE  raw_events_first.user_id = raw_events_second.user_id
            GROUP  BY raw_events_second.user_id
            HAVING SUM(raw_events_second.value_4) > 10) AS foo2 ) as f2
ON (f.id = f2.id)) as outer_most
GROUP BY
  outer_most.id;
```

We got the following vars associated with relations (i.e.., the member of common var eq class for this query)

```
INFO:  Relation Name: raw_events_first Var No:1
INFO:  Relation Name: raw_events_second Var No:2
INFO:  Relation Name: raw_events_first Var No:6
```

> varno is meaningless outside the Query it was in. Need to check how it used afterwards.

Do you still think in that way?